lib/vector-core/src/event/metadata.rs

```diff
@@ -213,6 +288,20 @@ impl EventMetadata {
         self
     }
 
+    /// Replaces the existing `source_type` with the given one.
+    #[must_use]
+    pub fn with_source_type(mut self, source_type: &'static str) -> Self {
+        self.source_type = Some(source_type);
+        self
+    }
+
+    /// Replaces the existing `DatadogMetricOriginMetadata` with the given one.
+    #[must_use]
+    pub fn with_origin_metadata(mut self, origin_metadata: DatadogMetricOriginMetadata) -> Self {
+        self.datadog_origin_metadata = Some(origin_metadata);
+        self
+    }
+
     /// Merge the other `EventMetadata` into this.
     /// If a Datadog API key is not set in `self`, the one from `other` will be used.
     /// If a Splunk HEC token is not set in `self`, the one from `other` will be used.
```

Comment:
Do we want to merge the origin metadata here?
This would impact what happens in the `aggregate` transform. If we are happy with always taking the origin from the first metric being aggregate then we can leave this.  There's probably no real perfect answer here, but we may prefer to override a `None` origin with a `Some` origin if we can.

Reply:
That is a good question.

I'm trying to think of a scenario where the metric series context (name, namespace, tags, etc.) , is the same for a set of metrics being aggregated, but the origin metadata is different.

If such a scenario could occur (which is not yet clear to me) , I would think we would not want to aggregate those metrics. Like the origin metadata would be considered part of the context. That way we maintain accuracy of the origin metadata.

Reply:
@neuronull I think it's more important for the aggregation to behave as expected than to retain all of the origin metadata. This would be a semantically breaking change that would (1) be unexpected to the user, (2) difficult to track down when there are inconsistencies in the downstream data store, and (3) uncontrollable by the end-user. Origin metadata is not part of the metric context in the same fashion as the metric name and tags, which are used to uniquely identify the metric. Partitioning by origin metadata in the aggregation could result in duplicate writes (two datapoints for the same timestamp) that are then merged in a last-write-wins fashion in the downstream data store, resulting in dropped data.

Reply:
>  If we are happy with always taking the origin from the first metric being aggregate then we can leave this.

I asked about the Agent's behavior in this type of scenario and the response received was

```
We definitely do not account for a scenario where a single metric series context has two different origin metadatas. We treat it similar to metric type which is we assume that it does not change for a given metric name.
```

So it seems OK to leave this as-is.